The ``101-lazy_matrix_mul`` module
=================================

Using ``lazy_matrix_mul``
--------------------------

Import ``lazy_matrix_mul`` from ``101-lazy_matrix_mul`` module
	>>> lazy_matrix_mul = __import__("101-lazy_matrix_mul").lazy_matrix_mul

Invalid arguments
	>>> lazy_matrix_mul()
	Traceback (most recent call last):
		...
	TypeError: lazy_matrix_mul() missing 2 required positional arguments: 'm_a' and 'm_b'

	>>> lazy_matrix_mul("a")
	Traceback (most recent call last):
		...
	TypeError: lazy_matrix_mul() missing 1 required positional argument: 'm_b'

	>>> lazy_matrix_mul("a", True)
	Traceback (most recent call last):
		...
	TypeError: m_a must be a list

	>>> lazy_matrix_mul([1, 2, 3], 234)
	Traceback (most recent call last):
		...
	TypeError: m_b must be a list

	>>> lazy_matrix_mul(None, True)
	Traceback (most recent call last):
		...
	TypeError: m_a must be a list

	>>> lazy_matrix_mul([1, 2, 3], None)
	Traceback (most recent call last):
		...
	TypeError: m_b must be a list

	>>> lazy_matrix_mul([1, 2, 3], [3, 2, 1])
	Traceback (most recent call last):
		...
	TypeError: m_a must be a list of lists

	>>> lazy_matrix_mul([[1], [2], [3]], [3, 2, 1])
	Traceback (most recent call last):
		...
	TypeError: m_b must be a list of lists

	>>> lazy_matrix_mul([[]], [[]])
	Traceback (most recent call last):
		...
	ValueError: shapes (1,0) and (1,0) not aligned: 0 (dim 1) != 1 (dim 0)

	>>> lazy_matrix_mul([[1]], [[]])
	array([], shape=(1, 0), dtype=float64)

	>>> lazy_matrix_mul([[{}, [23], (1,)]], [[1, 2, 3]])
	Traceback (most recent call last):
		...
	ValueError: shapes (1,3) and (1,3) not aligned: 3 (dim 1) != 1 (dim 0)

	>>> lazy_matrix_mul([[1]], [["a", True, 'A']])
	Traceback (most recent call last):
		...
	ValueError: data type must provide an itemsize
	
	>>> lazy_matrix_mul([[1, 2], [1]],  [[1], [1.3, 4]])
	Traceback (most recent call last):
		...
	TypeError: can't multiply sequence by non-int of type 'list'

	>>> lazy_matrix_mul([[1, 2], [4, 5]], [[1, 2], [1]])
	array([list([1, 2, 1, 1]), list([1, 2, 1, 2, 1, 2, 1, 2, 1, 1, 1, 1, 1])],
              dtype=object)

	>>> lazy_matrix_mul([[1, 2, 3]], [[1, 2], [3, 4]])
	Traceback (most recent call last):
		...
	ValueError: shapes (1,3) and (2,2) not aligned: 3 (dim 1) != 2 (dim 0)

Valid input
	>>> print(lazy_matrix_mul([[1, 2], [3, 4]], [[1, 2], [3, 4]]).tolist())
	[[7, 10], [15, 22]]
	
	>>> print(lazy_matrix_mul([[1, 2]], [[3, 4], [5, 6]]).tolist())
	[[13, 16]]

	>>> print(lazy_matrix_mul([[1, 2, 3], [4, 5, 6]], [[7, 8], [9, 10], [11, 12]]).tolist())
	[[58, 64], [139, 154]]
